File: src/animation.py
# src/animation.py

from PyQt5 import QtWidgets, QtGui, QtCore
import os
import sys

class GazeAnimation(QtWidgets.QGraphicsView):
    def __init__(self, eye_positions, frame_width, frame_height, fps=30, save_dir='frames'):
        super().__init__()
        self.eye_positions = eye_positions
        self.fps = fps
        self.current_frame = 0
        self.frame_width = frame_width
        self.frame_height = frame_height
        self.save_dir = save_dir

        # Create directory for frames if it doesn't exist
        os.makedirs(save_dir, exist_ok=True)

        # Set up the scene
        self.scene = QtWidgets.QGraphicsScene()
        self.setScene(self.scene)
        self.setFixedSize(frame_width, frame_height)

        # Calculate padding and eye dimensions
        x_values = [x for x, y in eye_positions if x is not None]
        y_values = [y for x, y in eye_positions if y is not None]
        
        min_x, max_x = min(x_values), max(x_values)
        min_y, max_y = min(y_values), max(y_values)
        padding_x = (max_x - min_x) * 0.1 if (max_x - min_x) != 0 else 1
        padding_y = (max_y - min_y) * 0.1 if (max_y - min_y) != 0 else 1

        ax_min_x, ax_max_x = min_x - padding_x, max_x + padding_x
        ax_min_y, ax_max_y = min_y - padding_y, max_y + padding_y
        self.setSceneRect(ax_min_x, ax_min_y, ax_max_x - ax_min_x, ax_max_y - ax_min_y)

        # Artificial eye setup
        center_x = (ax_min_x + ax_max_x) / 2
        center_y = (ax_min_y + ax_max_y) / 2
        radius = max((ax_max_x - ax_min_x), (ax_max_y - ax_min_y)) / 2 + max(padding_x, padding_y)
        
        # Draw the artificial eye
        self.eye = self.scene.addEllipse(
            center_x - radius, center_y - radius, 2 * radius, 2 * radius,
            pen=QtGui.QPen(QtCore.Qt.black, 2),
            brush=QtGui.QBrush(QtCore.Qt.white)
        )

        # Gaze point setup
        self.gaze_point = QtWidgets.QGraphicsEllipseItem(-7.5, -7.5, 15, 15)
        self.gaze_point.setBrush(QtGui.QBrush(QtCore.Qt.blue))
        self.scene.addItem(self.gaze_point)
        
        # Timer for animation
        self.timer = QtCore.QTimer()
        self.timer.timeout.connect(self.update_frame)
        self.timer.start(1000 / fps)

    def update_frame(self):
        if self.current_frame < len(self.eye_positions):
            x, y = self.eye_positions[self.current_frame]
            if x is not None and y is not None:
                self.gaze_point.setPos(x, y)
            else:
                self.gaze_point.setPos(self.scene.sceneRect().center())
            
            # Capture and save the current frame as an image
            self.capture_frame()

            self.current_frame += 1
        else:
            self.timer.stop()

    def capture_frame(self):
        # Capture the frame from the QGraphicsView
        image = self.grab()
        frame_path = os.path.join(self.save_dir, f'frame_{self.current_frame:04d}.png')
        image.save(frame_path, 'PNG')


def create_gaze_animation(eye_positions, frame_width, frame_height, fps=30, save_path='animation.mp4'):
    app = QtWidgets.QApplication(sys.argv)
    animation_view = GazeAnimation(eye_positions, frame_width, frame_height, fps)
    animation_view.show()

    app.exec_()

    # Instructions to compile the frames into a video using ffmpeg or similar tool
    print("\nFrames have been saved in the 'frames' directory.")
    print("To compile them into a video, you can use the following command in your terminal:")
    print(f"ffmpeg -r {fps} -i frames/frame_%04d.png -vcodec libx264 -y {save_path}")

File: src/animation.js
const { createCanvas } = require('canvas');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

class GazeAnimation {
    constructor(eyePositions, frameWidth, frameHeight, fps = 30, saveDir = 'frames') {
        this.eyePositions = eyePositions;
        this.fps = fps;
        this.currentFrame = 0;
        this.frameWidth = frameWidth;
        this.frameHeight = frameHeight;
        this.saveDir = saveDir;

        // Create directory for frames if it doesn't exist
        if (!fs.existsSync(saveDir)) {
            fs.mkdirSync(saveDir, { recursive: true });
        }

        // Setup canvas
        this.canvas = createCanvas(frameWidth, frameHeight);
        this.context = this.canvas.getContext('2d');

        // Calculate padding and eye dimensions
        const xValues = eyePositions.map(pos => pos[0]).filter(x => x !== null);
        const yValues = eyePositions.map(pos => pos[1]).filter(y => y !== null);

        const min_x = Math.min(...xValues);
        const max_x = Math.max(...xValues);
        const min_y = Math.min(...yValues);
        const max_y = Math.max(...yValues);

        const padding_x = (max_x - min_x) * 0.1 || 1;
        const padding_y = (max_y - min_y) * 0.1 || 1;

        this.ax_min_x = min_x - padding_x;
        this.ax_max_x = max_x + padding_x;
        this.ax_min_y = min_y - padding_y;
        this.ax_max_y = max_y + padding_y;

        // Artificial eye setup
        this.centerX = (this.ax_min_x + this.ax_max_x) / 2;
        this.centerY = (this.ax_min_y + this.ax_max_y) / 2;
        this.radius = Math.max((this.ax_max_x - this.ax_min_x), (this.ax_max_y - this.ax_min_y)) / 2 + Math.max(padding_x, padding_y);
    }

    drawEye() {
        // Draw the artificial eye
        this.context.beginPath();
        this.context.arc(this.centerX, this.centerY, this.radius, 0, Math.PI * 2);
        this.context.fillStyle = 'white';
        this.context.fill();
        this.context.lineWidth = 2;
        this.context.strokeStyle = 'black';
        this.context.stroke();
    }

    drawGazePoint(x, y) {
        // Draw gaze point
        this.context.beginPath();
        this.context.arc(x, y, 7.5, 0, Math.PI * 2);
        this.context.fillStyle = 'blue';
        this.context.fill();
    }

    updateFrame() {
        if (this.currentFrame < this.eyePositions.length) {
            const [x, y] = this.eyePositions[this.currentFrame];
            this.context.clearRect(0, 0, this.frameWidth, this.frameHeight);

            this.drawEye();

            if (x !== null && y !== null) {
                this.drawGazePoint(x, y);
            } else {
                this.drawGazePoint(this.frameWidth / 2, this.frameHeight / 2);
            }

            this.captureFrame();
            this.currentFrame += 1;

            // Schedule next frame
            setTimeout(() => this.updateFrame(), 1000 / this.fps);
        } else {
            console.log("Animation complete.");
        }
    }

    captureFrame() {
        // Save current frame as an image
        const framePath = path.join(this.saveDir, `frame_${String(this.currentFrame).padStart(4, '0')}.png`);
        const buffer = this.canvas.toBuffer('image/png');
        fs.writeFileSync(framePath, buffer);
        console.log(`Saved frame: ${framePath}`);
    }
}

// Main function to create the animation
function createGazeAnimation(eyePositions, frameWidth, frameHeight, fps = 30, savePath = 'animation.mp4') {
    const animation = new GazeAnimation(eyePositions, frameWidth, frameHeight, fps);
    animation.updateFrame();

    // Compile frames into video with FFmpeg
    const ffmpegCommand = `ffmpeg -r ${fps} -i ${animation.saveDir}/frame_%04d.png -vcodec libx264 -y ${savePath}`;
    console.log("Frames saved. To compile them into a video, run the following command:");
    console.log(ffmpegCommand);

    // Optionally, execute FFmpeg directly in Node.js
    exec(ffmpegCommand, (error, stdout, stderr) => {
        if (error) {
            console.error(`Error during video creation: ${error.message}`);
            return;
        }
        console.log(`Video saved to ${savePath}`);
    });
}

// Example usage
const exampleEyePositions = [
    [50, 50], [60, 60], [70, 80], [80, 90], [90, 100], // Add more coordinates as needed
];
createGazeAnimation(exampleEyePositions, 640, 480, 30);
